samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -16,173 +16,109 @@
+   * @param <WM>  the type of {@link WindowOutput} {@link Message} from the {@link Window} function
+   * @return  the output {@link MessageStream} after applying the window function on the input {@link MessageStream}
    */
-  public <WK, WV, WS extends WindowState<WV>, WM extends WindowOutput<WK, WV>> MessageStream<WM> window(Windows.Window<M, WK, WV, WM> window) {
-    Operator<WM> wndOp = Operators.getWindowOperator(Windows.getInternalWindowFn(window));
-    this.subscribers.add(wndOp);
-    return wndOp.getOutputStream();
-  }
+  <WK, WV, WS extends WindowState<WV>, WM extends WindowOutput<WK, WV>> MessageStream<WM> window(
+      Window<M, WK, WV, WM> window);
 
   /**
-   * Method to add an input {@link MessageStream} to a join function. Note that we currently only support 2-way joins.
+   * Joins this {@link MessageStream} with another {@link MessageStream} using the provided pairwise {@link JoinFunction}.
+   *
+   * We currently only support 2-way joins.
    *
-   * @param other  the other stream to be joined w/
-   * @param merger  the common function to merge messages from this {@link MessageStream} and {@code other}
+   * @param otherStream  the other stream to be joined with
+   * @param joinFn  the function to join {@link Message}s from this and the other {@link MessageStream}
    * @param <K>  the type of join key
-   * @param <JM>  the type of message in the {@link Message} from the other join stream
-   * @param <RM>  the type of message in the {@link Message} from the join function
-   * @return the output {@link MessageStream} from the join function {@code joiner}
+   * @param <JM>  the type of {@link Message}s in the other join stream
+   * @param <RM>  the type of {@link Message} resulting from the join
+   * @return  the output {@link MessageStream} after applying the join function to the two {@link MessageStream}
    */
-  public <K, JM extends Message<K, ?>, RM extends Message> MessageStream<RM> join(MessageStream<JM> other,
-      BiFunction<M, JM, RM> merger) {
-    MessageStream<RM> outputStream = new MessageStream<>();
-
-    BiFunction<M, JM, RM> parJoin1 = merger::apply;
-    BiFunction<JM, M, RM> parJoin2 = (m, t1) -> merger.apply(t1, m);
-
-    // TODO: need to add default store functions for the two partial join functions
-
-    other.subscribers.add(Operators.<JM, K, M, RM>getPartialJoinOperator(parJoin2, outputStream));
-    this.subscribers.add(Operators.<M, K, JM, RM>getPartialJoinOperator(parJoin1, outputStream));
-    return outputStream;
-  }
+  <K, JM extends Message<K, ?>, RM extends Message> MessageStream<RM> join(MessageStream<JM> otherStream,
+      JoinFunction<M, JM, RM> joinFn);
 
   /**
-   * Method to merge all {@code others} streams w/ this {@link MessageStream}. The merging streams must have the same type {@code M}
+   * Merge all {@code otherStreams} with this {@link MessageStream}.
+   * The merging streams must have the same {@link Message} type {@code M}.
```

Comment:
Is there any `ordering` among messages in the merged steram? If so, Should we call it out?